proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,285 @@
+#![deny(missing_docs)]
+use control::pb::telemetry::{
+    Latency as LatencyProto,
+};
+
+use std::{ops, slice, u32};
+use std::cell::Cell;
+use std::default::Default;
+use std::time::Duration;
+
+/// A series of latency values and counts.
+#[derive(Debug)]
+pub struct Latencies(Buckets<Latency, Buckets<Latency>>);
+
+/// A latency in tenths of a millisecond.
+#[derive(Debug, Default, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Hash)]
+pub struct Latency(u32);
+
+/// Generalization of a histogram bucket counting observations of type `T`.
+trait Bucket<T: PartialOrd> {
+    /// Returns `true` if a given `observation` falls into this `Bucket`.
+    fn contains(&self, observation: &T) -> bool;
+    /// Observe a value, incrementing this bucket if it `contains` that value.
+    /// 
+    /// # Returns
+    /// - `true` if this `Bucket`'s count was incremented;
+    /// - `false` otherwise.
+    fn add(&self, observation: &T) -> bool;
+}
+
+/// A "leaf" bucket containing a single counter.
+/// 
+/// # Type Parameters:
+/// - `T`: the type of observations counted by this bucket.
+/// - `C`: the type used for counting observed values. Bounded by `Copy`
+///        as it is placed in a `Cell`; defaults to `u32`.
+#[derive(Debug, Clone)]
+struct LeafBucket<T, C: Copy = u32> {
+    /// The upper bound of the range of values in this bucket.
+    max: T,
+    /// Count of observations falling into this bucket.
+    /// 
+    /// `Cell` is used for interior mutability here, as it should be possible
+    /// to increment the count in a bucket without mutably locking the entire 
+    /// histogram structure containing this `LeafBucket`; as would be necessary to
+    /// change the _shape_ of the histogram.
```

Comment:
This probably shouldn't be a `Cell`, and just use normal mutability to mutate this.

Reply:
Okay, that's fair. I suspect a situation where we're trying to mutate the value in multiple buckets at once is pretty unlikely with the current implementation anyway...